File: day13/main.py

import functools
import json

def cmp(a1, a2):
    if isinstance(a1, int) and isinstance(a2, int):
        return a1 - a2
    elif isinstance(a1, int):
        return cmp([a1], a2)
    elif isinstance(a2, int):
        return cmp(a1, [a2])
    else:
        # both lists
        i = 0
        while i < len(a1) and i < len(a2):
            x = cmp(a1[i], a2[i])
            if x != 0:
                return x
            i += 1
        return len(a1) - len(a2)


def part1(lines):
    idx = 1
    sum = 0
    i = 0
    while i < len(lines):
        line1 = lines[i].strip()
        line2 = lines[i+1].strip()
        i += 3
        
        if cmp(json.loads(line1), json.loads(line2)) <= 0:
            sum += idx
            print(str(idx) + " is sorted")
        else:
            print(str(idx) + " invalid")
        idx += 1
    print(sum)

def part2(lines):
    i = 0

    div1 = [[2]]
    div2 = [[6]]

    outLines = [div1,div2]

    while i < len(lines):
        line1 = lines[i].strip()
        line2 = lines[i+1].strip()
        i += 3
        outLines.append(json.loads(line1))
        outLines.append(json.loads(line2))
    outLines.sort(key=functools.cmp_to_key(cmp))
        
    print((outLines.index(div1) + 1) * (outLines.index(div2) + 1))
        
        

with open("../inputs/day13.txt","r") as f:
    lines = f.readlines()
    part1(lines)
with open("../inputs/day13.txt","r") as f:
    lines = f.readlines()
    part2(lines)

testInput = """
[1,1,3,1,1]
[1,1,5,1,1]

[[1],[2,3,4]]
[[1],4]

[9]
[[8,7,6]]

[[4,4],4,4]
[[4,4],4,4,4]

[7,7,7,7]
[7,7,7]

[]
[3]

[[[]]]
[[]]

[1,[2,[3,[4,[5,6,7]]]],8,9]
[1,[2,[3,[4,[5,6,0]]]],8,9]
""".strip().split("\n")
part1(testInput)
part2(testInput)

File: day13/main.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"sort"
	"strings"

	"github.com/strangelytyped/advent-of-code-2022/utils"
)

type Token struct {
	intVal   int
	children []*Token
	isInt    bool
}

func wrap(a1 *Token) *Token {
	return &Token{children: []*Token{a1}}
}

func cmp(a1, a2 *Token) int {
	if a1.isInt && a2.isInt {
		return a1.intVal - a2.intVal
	} else if a1.isInt {
		return cmp(wrap(a1), a2)
	} else if a2.isInt {
		return cmp(a1, wrap(a2))
	} else {
		// both lists
		for i := 0; i < len(a1.children) && i < len(a2.children); i++ {
			x := cmp(a1.children[i], a2.children[i])
			if x != 0 {
				return x
			}
		}
		return len(a1.children) - len(a2.children)
	}
}

func parse(line string, token *Token) int {
	idx := 0
	if line[0] == '[' {
		// Array mode
		idx++
		for idx < len(line) {
			childToken := Token{}
			token.children = append(token.children, &childToken)
			readChrs := parse(line[idx:], &childToken)
			idx += readChrs
			if line[idx] == ']' {
				idx++
				break
			}
			idx++
		}
	} else if line[0] == ']' {
		return 0
	} else {
		// number mode
		val := 0
		for {
			if line[idx] == ',' || line[idx] == ']' {
				break
			}
			val = val*10 + int(line[idx]-'0')
			idx++
		}
		token.intVal = val
		token.isInt = true
	}
	return idx
}

func Part1(input io.Reader) int {
	scanner := bufio.NewScanner(input)
	idx := 1
	sum := 0
	for scanner.Scan() {
		line1 := scanner.Text()
		scanner.Scan()
		line2 := scanner.Text()
		scanner.Scan()

		root1 := Token{}
		parse(line1, &root1)
		root2 := Token{}
		parse(line2, &root2)

		if cmp(&root1, &root2) <= 0 {
			fmt.Printf("Match %d\n", idx)
			sum += idx
		} else {
			fmt.Printf("Fail %d\n", idx)
		}
		idx++
	}
	return sum
}

func makeIntToken(x int) *Token {
	return &Token{intVal: x, isInt: true}
}

func Part2(input io.Reader) int {
	scanner := bufio.NewScanner(input)

	tokenA := wrap(makeIntToken(2))
	tokenB := wrap(makeIntToken(6))

	lines := []*Token{tokenA, tokenB}

	for scanner.Scan() {
		line1 := scanner.Text()
		if len(strings.TrimSpace(line1)) == 0 {
			continue
		}

		root1 := Token{}
		parse(line1, &root1)
		lines = append(lines, &root1)
	}

	sort.Slice(lines, func(i, j int) bool {
		return cmp(lines[i], lines[j]) < 0
	})
	return (index(lines, tokenA) + 1) * (index(lines, tokenB) + 1)
}

func index(lines []*Token, tokenA *Token) int {
	for i, x := range lines {
		if tokenA == x {
			return i
		}
	}
	return -1
}

func main() {
	utils.Run("day13.txt", Part1, Part2)
}
